nvflare/edge/assessors/buff_device_manager.py

```diff
@@ -66,17 +178,30 @@ def fill_selection(self, current_model_version: int, fl_ctx) -> None:
             usable_devices = set(self.available_devices.keys()) - set(self.used_devices.keys())
 
             if usable_devices:
-                for _ in range(num_holes):
-                    device_id = random.choice(list(usable_devices))
-                    usable_devices.remove(device_id)
-                    # current_selection keeps track of devices selected for a particular model version
-                    self.current_selection[device_id] = current_model_version
-                    self.used_devices[device_id] = {
-                        "model_version": current_model_version,
-                        "selection_version": self.current_selection_version,
-                    }
-                    if not usable_devices:
-                        break
+                if self.device_sampling_strategy == "balanced":
+                    # try to balance the usage of devices across clients
+                    selected_devices = self._balanced_device_sampling(usable_devices, num_holes)
+                    for device_id in selected_devices:
+                        # current_selection keeps track of devices selected for a particular model version
+                        self.current_selection[device_id] = current_model_version
+                        self.used_devices[device_id] = {
+                            "model_version": current_model_version,
+                            "selection_version": self.current_selection_version,
+                        }
+                elif self.device_sampling_strategy == "random":
+                    for _ in range(num_holes):
+                        device_id = random.choice(list(usable_devices))
+                        usable_devices.remove(device_id)
+                        # current_selection keeps track of devices selected for a particular model version
+                        self.current_selection[device_id] = current_model_version
+                        self.used_devices[device_id] = {
+                            "model_version": current_model_version,
+                            "selection_version": self.current_selection_version,
+                        }
+                        if not usable_devices:
+                            break
+                else:
+                    raise ValueError(f"Invalid device sampling strategy: {self.device_sampling_strategy}")
```

Comment:
can this sampling strategy be a class they can customize, and we just pass an object in?
Should we introduce another concept "SampleStrategy" or something for the edge stuff?